### PR TITLE
Feature / HOLO 1396 Add BridgeCollection Class to SDK

### PR DESCRIPTION
**Pull request by BravoNatalie**
## Describe Changes

Changes made to the package: **sdk**.

I made it better by doing ...

- Add `simulateContract` functions to `HolographBaseContract`, this functions should be use for `staticCall`.
- Introduce a base class, `BridgeAsset`, to serve as the foundation for all specific bridge classes. 
- Create `BridgeCollection` class
- Create `utils/gas` containing functions to estimate gas costs. This initial implementation copies functionality from the frontend code. It needs further improvements.
- Create gas controller for bridge collections
- Add minor improvements

## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [x] Code styles have been enforced
- [x] All tests are passing

### CODE REVIEW

File: packages/sdk/src/assets/bridge-collection.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+    const initCode = encodeAbiParameters(
+      parseAbiParameters([
+        'DeploymentConfig config',
+        'struct DeploymentConfig { bytes32 contractType; uint32 chainType; bytes32 salt; bytes byteCode; bytes initCode; }',
+        'Verification signature',
+        'struct Verification { bytes32 r; bytes32 s; uint8 v; }',
+        'address signer',
+      ]),
+      // @ts-ignore NOTE: for some reason the inference for the the return type is wrong
```
Comment (feliperodriguess) [inline packages/sdk/src/assets/bridge-collection.ts:43]: This is not making so much sense, there are 5 parameters in the first argument array, and only 3 in the array of values. The length between those arrays should match, like 5 ABI parameters and 5 actual values

File: packages/sdk/src/assets/collection-legacy.ts [hunk 2]
```diff
@@ -9,114 +9,108 @@ import {decodeBridgeableContractDeployedEvent} from '../utils/decoders'
 import {allEventsEnabled, destructSignature, generateRandomSalt, parseBytes} from '../utils/helpers'
 import {evm2hlg, remove0x} from '../utils/transformers'
 import {Erc721Config, GasFee, HolographConfig, Signature, SignDeploy, WriteContractOptions} from '../utils/types'
+import {create2AddressFromDeploymentHash, getErc721DeploymentConfigHash} from '../utils/encoders'
 
 export class HolographLegacyCollection {
-  collectionInfo: CollectionInfo
-  holographConfig: HolographConfig
```
Comment (feliperodriguess) [inline packages/sdk/src/assets/collection-legacy.ts:15]: Please add this line back, this is really important when we deal with the NFT classes

File: packages/sdk/src/assets/collection-legacy.ts [hunk 3]
```diff
@@ -9,114 +9,108 @@ import {decodeBridgeableContractDeployedEvent} from '../utils/decoders'
...
+  public signature?: Signature
+  public txHash?: string
+
+  private readonly factory: Factory
+  private readonly registry: Registry
+
+  constructor(public holographConfig: HolographConfig, {collectionInfo, primaryChainId}: CreateLegacyCollection) {
+    this._collectionInfo = validate.collectionInfo.parse(collectionInfo)
     this.primaryChainId = validate.primaryChainId.parse(primaryChainId)
-    this.holographConfig = configObject
```
Comment (feliperodriguess) [inline packages/sdk/src/assets/collection-legacy.ts:31]: please add this back

**Review by feliperodriguess — CHANGES_REQUESTED** [review]

File: packages/sdk/src/contracts/holograph-base.contract.ts [hunk 4]
```diff
@@ -61,6 +69,21 @@ export class HolographBaseContract {
...
+    functionName,
+    args,
+    options,
+  }: SimulateContractArgs<TAbi>) {
+    const provider = this._providers.byChainId(chainId)
+
+    let client = {public: provider}
+    const contract: GetContractReturnType<Abi, typeof client> = getContract({address, abi: this._abi, client})
+
+    return await contract.simulate[functionName](args as unknown[], (options as SimulateContractParameters) || {})
```
Comment (feliperodriguess) [inline packages/sdk/src/contracts/holograph-base.contract.ts:84]: you can remove the `await`

File: packages/sdk/src/assets/bridge-asset.ts [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+      bridgeOutPayload,
+    )
+
+    const holographToChainId = getNetworkByChainId(toChainId).holographId
+
+    // now that we have gasLimit and gasPrice, we update the payload to include the proper data needed for estimating fees
+    const bridgeRequestPayload = (
+      await this._createBridgeOutPayload(
+        chainId,
+        holographToChainId,
```
Comment (sogoiii) [inline packages/sdk/src/assets/bridge-asset.ts:182]: the name makes me think that its doing a conversion. Maybe  `destinationHolographChainId`. that way it keeps withe same form as the other destination variables.

File: packages/sdk/src/utils/helpers.ts [hunk 6]
```diff
@@ -4,6 +4,7 @@ import {NETWORK_KEY_BY_RPC_URL, NetworkKey, networks} from '@holographxyz/networ
 
 import {getEnv} from '../config/env.validation'
 import {HolographConfig, NetworkRpc, Signature} from './types'
+import {Address} from 'abitype'
```
Comment (feliperodriguess) [inline packages/sdk/src/utils/helpers.ts:7]: This was imported and not used

File: packages/sdk/src/utils/types.ts [hunk 7]
```diff
@@ -40,29 +41,35 @@ export type BridgeInErc721Args = {
...
+export type Signature = {
+  readonly r: Hex
+  s: Hex
+  v: Hex | number
 }
 
-export type Signature = DeploymentConfig['signature']
+export type DeploymentConfigSettings = {
+  readonly config: DeploymentConfig
+
```
Comment (feliperodriguess) [inline packages/sdk/src/utils/types.ts:64]: you can remove this line

File: packages/sdk/src/utils/types.ts [hunk 8]
```diff
@@ -40,29 +41,35 @@ export type BridgeInErc721Args = {
...
+  readonly signature: Signature
+
```
Comment (feliperodriguess) [inline packages/sdk/src/utils/types.ts:66]: this one too

File: packages/sdk/src/assets/bridge-asset.ts [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+      bridgeRequestPayload,
+    )
+
+    const unsignedBridgeOutTx = BridgeAsset.createUnsignedBridgeOutTx(
+      holographToChainId,
+      contractAddress,
+      destinationGasLimit,
+      destinationGasPrice,
+      bridgeOutPayload,
+      //{value: feeValue}, TODO: do we need to pass the value here?
```
Comment (sogoiii) [inline packages/sdk/src/assets/bridge-asset.ts:205]: I believe it should be here. As this should generate the entire tx payload, except that its not singed and submitted.

File: packages/sdk/src/utils/gas.ts [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+      if (typeof block.transactions[0] !== 'string') {
+        // we have a Block with transactions
+        for (let i = 0, l = block.transactions.length; i < l; i++) {
+          const tx = block.transactions[i] as Transaction
+          if (gasPricing.isEip1559) {
+            // set current tx priority fee
+            let priorityFee: bigint = BigInt(0)
+            let remainder: bigint
+
+            if (tx.maxFeePerGas === undefined || tx.maxPriorityFeePerGas === undefined) {
```
Comment (feliperodriguess) [inline packages/sdk/src/utils/gas.ts:132]: We can certainly improve this logic, there are a bunch of ifs/elses, but I guess that could be with another PR

File: packages/sdk/src/assets/bridge-asset.ts [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+      chain: walletClient.chain,
+      account: wallet.account,
+      data: unsignedTx,
+      gasPrice: options?.sourceGasPrice ?? gasSource.gasPrice,
+      gas: options?.sourceGasLimit ?? gasSource.gasLimit,
+      value,
+    })
+
+    //@ts-ignore TODO: need to fix the holographWallet _multiChainWalletClient type, since the created WalletClient extend publicActions
+    const receipt = await wallet.waitForTransactionReceipt({hash: tx}) // TODO: should we wait for the transaction to be processed?
```
Comment (sogoiii) [inline packages/sdk/src/assets/bridge-asset.ts:264]: Maybe we should make this an option? Something like `option.shouldWait`? by default its set to false, but if set to true this function will hold.

File: packages/sdk/src/assets/bridge-collection.ts [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+  }
+
+  async getInitCode() {
+    if (!this._initCode) {
+      this._initCode = await BridgeCollection.createInitCode(this.chainId, this.erc721DeploymentConfig, this.wallet)
+    }
+    return this._initCode
+  }
+
+  async bridgeOut(toChainId: number): Promise<TransactionReceipt> {
```
Comment (feliperodriguess) [inline packages/sdk/src/assets/bridge-collection.ts:57]: I believe this should also accept options param (gas overrides) as we have in the deploy and mint methods from collection & nft asset classes. 

And we also should have a test file simulating a collection bridge action right?

File: packages/sdk/src/assets/collection-moe.ts [hunk 13]
```diff
@@ -206,30 +205,31 @@ export class HolographMoeERC721DropV1 {
...
-  _getMetadataRendererAddress(chainId = this.primaryChainId) {
+  protected _getMetadataRendererAddress(chainId = this.primaryChainId) {
     return Addresses.editionsMetadataRendererV1()
   }
 
-  _getDropContractType() {
-    return '0x' + stringToHex('HolographDropERC721').substring(2).padStart(64, '0')
+  protected _getDropContractType() {
+    return parseBytes('HolographDropERC721')
+    //'0x' + stringToHex('HolographDropERC721').substring(2).padStart(64, '0')
```
Comment (sogoiii) [inline packages/sdk/src/assets/collection-moe.ts:221]: Remove this commented line.

File: packages/sdk/src/assets/collection-moe.ts [hunk 14]
```diff
@@ -473,19 +476,20 @@ export class HolographMoeERC721DropV2 extends HolographMoeERC721DropV1 {
...
-  _getMetadataRendererAddress(chainId = this.primaryChainId) {
+  protected _getMetadataRendererAddress(chainId = this.primaryChainId) {
     return Addresses.editionsMetadataRenderer(getEnv().HOLOGRAPH_ENVIRONMENT, chainId)
   }
 
-  _getDropContractType() {
-    return '0x' + stringToHex('HolographDropERC721V2').substring(2).padStart(64, '0')
+  protected _getDropContractType() {
+    return parseBytes('HolographDropERC721V2')
+    //return '0x' + stringToHex('HolographDropERC721V2').substring(2).padStart(64, '0')
```
Comment (sogoiii) [inline packages/sdk/src/assets/collection-moe.ts:485]: remove commented out line

File: packages/sdk/src/assets/bridge-asset.ts [hunk 15]
```diff
@@ -0,0 +1,267 @@
...
+        bridgeOutPayload,
+      )
+
+      const consumedGas = BigInt(
+        (
+          await this._operator.simulateContractFunction({
+            chainId,
+            functionName: 'jobEstimator',
+            args: [bridgeRequestPayloadStub],
+            options: {gas: gasLimitStub}, //{account: gasEstimationHelperAddress , gas: TEST_GAS_LIMIT},
```
Comment (sogoiii) [inline packages/sdk/src/assets/bridge-asset.ts:130]: Is there a reason we are not using the `gasEstimationHelperAddress` variable from above?

**Review by sogoiii — CHANGES_REQUESTED** [review]

File: packages/sdk/src/utils/gas.ts [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+      destinationFrom = '0x0000000000000000000000000000000000000000'
+      break
+  }
+
+  return destinationFrom
+}
+
+// This function is here to accomodate instances where a network has a minimum BaseBlockFee
+export function adjustBaseBlockFee(chainId: number, baseBlockFee: bigint): bigint {
+  const network: string = getNetworkByChainId(chainId).key
```
Comment (sogoiii) [inline packages/sdk/src/utils/gas.ts:61]: Can you add fuji to this case. And set it as 3 Gwei for binance and binance testnet

File: packages/sdk/src/assets/bridge-asset.ts [hunk 17]
```diff
@@ -0,0 +1,267 @@
...
+    const {unsignedTx, gasSource, value} = await this.estimateBridgeOut(
+      chainId,
+      toChainId,
+      contractAddress,
+      bridgeOutPayload,
+    )
+
+    const walletClient = wallet.onChain(chainId)
+
+    const tx = await walletClient.sendTransaction({
```
Comment (feliperodriguess) [inline packages/sdk/src/assets/bridge-asset.ts:254]: This approach is different from what we've been doing lately.

We've been using the writeContract method from HolographBase instead of  `walletClient.sendTransaction`. Let's choose only one option for consistency.

We can also pass overrides to the writeContract methods, such as gas, gasPrice, value etc. You can take a look at how we invoked functions such as `mint`, `deploy` etc

File: packages/sdk/src/assets/bridge-asset.ts (hunk 11 above)
Comment (feliperodriguess) [reply packages/sdk/src/assets/bridge-asset.ts:264]: It depends more in what is the goal of the function/tx

On the collection and nft classes (mostly deploying and minting functions), we have to wait because we need to get the tokenId, collection address, and things like that from the receipt. So waiting for the tx is mandatory to decode its receipt

We need to ask ourselves, do we need the tx receipt for the return of `_bridgeOut` function? If not, then we can remove this line and only return the txHash

Comment (feliperodriguess) [reply packages/sdk/src/assets/bridge-asset.ts:264]: Maybe we don't need an argument for that. It depends more in what is the goal of the function/tx

On the collection and nft classes (mostly deploying and minting functions), we have to wait because we need to get the tokenId, collection address, and things like that from the receipt. So waiting for the tx is mandatory to decode its receipt

We need to ask ourselves, do we need the tx receipt for the return of `_bridgeOut` function? If not, then we can remove this line and only return the txHash

Comment (feliperodriguess) [reply packages/sdk/src/assets/bridge-asset.ts:264]: @sogoiii  Maybe we don't need an argument for that. It depends more in what is the goal of the function/tx

On the collection and nft classes (mostly deploying and minting functions), we have to wait because we need to get the tokenId, collection address, and things like that from the receipt. So waiting for the tx in those cases is mandatory to decode its receipt

We need to ask ourselves, do we need the tx receipt for the return of `_bridgeOut` function? If not, then we can remove this line and only return the txHash

File: packages/sdk/src/assets/bridge-collection.ts (hunk 1 above)
Comment (BravoNatalie) [reply packages/sdk/src/assets/bridge-collection.ts:43]: The parameters are correct. There are 5 arguments because 2 of them are the structs definition. So, in total, there are only 3 arguments. I tested this function, and it works as expected, but the inference is wrong.

File: packages/sdk/src/assets/bridge-asset.ts (hunk 5 above)
Comment (BravoNatalie) [reply packages/sdk/src/assets/bridge-asset.ts:182]: good idea

File: packages/sdk/src/assets/bridge-asset.ts (hunk 15 above)
Comment (BravoNatalie) [reply packages/sdk/src/assets/bridge-asset.ts:130]: It's giving me TypeScript problems, even though it's supposed to accept this parameter. I'm going to try a bit more. If I'm not able to pass this value, I'll add a @ts-ignore.

File: packages/sdk/src/utils/gas.ts (hunk 16 above)
Comment (BravoNatalie) [reply packages/sdk/src/utils/gas.ts:61]: isn't avalancheTestnet the same as fuji?

File: packages/sdk/src/assets/bridge-collection.ts (hunk 1 above)
Comment (feliperodriguess) [reply packages/sdk/src/assets/bridge-collection.ts:43]: maybe we can remove the struct definitions then

File: packages/sdk/src/assets/bridge-asset.ts (hunk 17 above)
Comment (feliperodriguess) [reply packages/sdk/src/assets/bridge-asset.ts:254]: Ah ok, because of the `data` argument right, I think using writeContract we don't support this argument. So it's all good. 

Then let's use this approach only in this case, and keep using the writeContract as the "default" approach

File: packages/sdk/src/assets/bridge-asset.ts (hunk 11 above)
Comment (sogoiii) [reply packages/sdk/src/assets/bridge-asset.ts:264]: @feliperodriguess @BravoNatalie ok, lets make it wait for the tx to do some of the decoding. We can add options later.

File: packages/sdk/src/utils/gas.ts (hunk 16 above)
Comment (sogoiii) [reply packages/sdk/src/utils/gas.ts:61]: @BravoNatalie no you are right. `avalancheTestnet` is fine. `fuji` is the short name version.

File: packages/sdk/src/tests/assets/bridge-asset.test.ts [hunk 18]
```diff
@@ -0,0 +1,202 @@
...
+  describe.skip('_bridgeOut()', () => {
+    it('should send a bridge out resquest transaction', async () => {
+      const tx = await bridgeAsset['_bridgeOut'](
+        sourceChainId,
+        destinationChainId,
+        holographableContract,
+        bridgeOutPayload,
+        wallet,
+      )
+      expect(tx)
```
Comment (feliperodriguess) [inline packages/sdk/src/tests/assets/bridge-asset.test.ts:199]: This seems incomplete

File: packages/sdk/src/tests/assets/collection-legacy.test.ts [hunk 19]
```diff
@@ -1,12 +1,13 @@
...
+import {beforeAll, beforeEach, describe, expect, it} from 'vitest'
+import {Address, isAddress} from 'viem'
 
 import {HolographLegacyCollection} from '../../assets/collection-legacy'
 import {ContractRevertError} from '../../errors'
 import {HolographWallet} from '../../services'
 import {configObject, localhostContractAddresses, LOCALHOST2_CHAIN_ID} from '../setup'
 import {generateRandomSalt, sleep} from '../../utils/helpers'
 import {HolographAccount} from '../../utils/types'
+import {getErc721DeploymentConfigHash} from '../../utils/encoders'
```
Comment (feliperodriguess) [inline packages/sdk/src/tests/assets/collection-legacy.test.ts:10]: This was imported and not used

File: packages/sdk/src/tests/assets/collection-legacy.test.ts [hunk 20]
```diff
@@ -1,12 +1,13 @@
-import {beforeEach, describe, expect, it} from 'vitest'
-import {isAddress} from 'viem'
+import {beforeAll, beforeEach, describe, expect, it} from 'vitest'
```
Comment (feliperodriguess) [inline packages/sdk/src/tests/assets/collection-legacy.test.ts:1]: `beforeAll` was imported and not used

File: packages/sdk/src/tests/assets/collection-legacy.test.ts [hunk 21]
```diff
@@ -1,12 +1,13 @@
...
+import {Address, isAddress} from 'viem'
```
Comment (feliperodriguess) [inline packages/sdk/src/tests/assets/collection-legacy.test.ts:2]: `Address` was imported and not used

File: packages/sdk/src/utils/gas.ts [hunk 22]
```diff
@@ -0,0 +1,191 @@
+import {getNetworkByChainId, networks} from '@holographxyz/networks'
+import {Address, Block, Hex, PublicClient, Transaction, hexToBigInt} from 'viem'
+import {GasPricing} from './types'
```
Comment (feliperodriguess) [inline packages/sdk/src/utils/gas.ts:3]: break line here

**Review by feliperodriguess — CHANGES_REQUESTED** [review]

File: packages/sdk/src/assets/bridge-asset.ts [hunk 23]
```diff
@@ -0,0 +1,271 @@
...
+      chain: walletClient.chain,
+      account: wallet.account,
+      data: unsignedTx,
+      gasPrice: options?.sourceGasPrice ?? gasSource.gasPrice,
+      gas: options?.sourceGasLimit ?? gasSource.gasLimit,
+      value,
+    })
+
+    //@ts-ignore TODO: need to fix the holographWallet _multiChainWalletClient type, since the created WalletClient extend publicActions
+    return walletClient.getTransaction({hash: tx})
```
[the remark this answers is not in the document]
Comment (feliperodriguess) [reply packages/sdk/src/assets/bridge-asset.ts:334]: Got it, ok

**Review by feliperodriguess — APPROVED** [review]

**Review by sogoiii — APPROVED** [review]
👍
